File: spark-scala/src/main/scala/io/kore/spark/KoreTable.scala
package io.kore.spark

import org.apache.spark.sql.connector.read.{Scan, ScanBuilder}
import org.apache.spark.sql.connector.write.{Write, WriteBuilder}
import org.apache.spark.sql.connector.catalog.Table
import org.apache.spark.sql.types.StructType
import org.apache.spark.sql.util.CaseInsensitiveStringMap

/**
 * Kore Table Implementation
 * 
 * Handles read and write operations for Kore files
 */
class KoreTable(
    val path: String,
    val schema: StructType,
    val options: CaseInsensitiveStringMap) extends Table {

  override def name(): String = s"kore@$path"

  override def schema(): StructType = schema

  override def capabilities(): java.util.Set[org.apache.spark.sql.connector.catalog.TableCapability] = {
    // TODO: Define capabilities (read, write, streaming, etc.)
    java.util.EnumSet.noneOf(classOf[org.apache.spark.sql.connector.catalog.TableCapability])
  }

  override def newScanBuilder(options: CaseInsensitiveStringMap): ScanBuilder = {
    // TODO: Create KoreScanBuilder for read operations
    throw new RuntimeException("Not yet implemented")
  }

  override def newWriteBuilder(options: CaseInsensitiveStringMap): WriteBuilder = {
    // TODO: Create KoreWriteBuilder for write operations
    throw new RuntimeException("Not yet implemented")
  }
}

File: maven/src/main/java/com/github/arunkatherashala/kore/spark/KoreTable.java
package com.github.arunkatherashala.kore.spark;

import org.apache.spark.sql.connector.catalog.SupportsRead;
import org.apache.spark.sql.connector.catalog.SupportsWrite;
import org.apache.spark.sql.connector.catalog.Table;
import org.apache.spark.sql.connector.catalog.TableCapability;
import org.apache.spark.sql.connector.read.ScanBuilder;
import org.apache.spark.sql.connector.write.LogicalWriteInfo;
import org.apache.spark.sql.connector.write.WriteBuilder;
import org.apache.spark.sql.types.StructType;
import org.apache.spark.sql.util.CaseInsensitiveStringMap;

import java.util.HashSet;
import java.util.Set;

public class KoreTable implements Table, SupportsRead, SupportsWrite {

    private final String path;
    private final StructType schema;

    public KoreTable(String path, StructType schema) {
        this.path = path;
        this.schema = schema;
    }

    @Override
    public String name() { return "kore:" + path; }

    @Override
    public StructType schema() { return schema; }

    @Override
    public Set<TableCapability> capabilities() {
        Set<TableCapability> caps = new HashSet<>();
        caps.add(TableCapability.BATCH_READ);
        caps.add(TableCapability.BATCH_WRITE);
        caps.add(TableCapability.TRUNCATE);
        return caps;
    }

    @Override
    public ScanBuilder newScanBuilder(CaseInsensitiveStringMap options) {
        return new KoreScanBuilder(path, schema);
    }

    @Override
    public WriteBuilder newWriteBuilder(LogicalWriteInfo info) {
        return new KoreWriteBuilder(path, info.schema());
    }
}
